src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.cpp

```diff
@@ -529,12 +530,284 @@ void MultiDiracDeterminant::mw_accept_rejectMove(const RefVectorWithLeader<Multi
+
+  for (int iw = 0, iacc = 0, irej = 0; iw < nw; iw++)
     if (isAccepted[iw])
-      wfc_list[iw].acceptMove(p_list[iw], iat, false);
+      idx_Accepted[iacc++] = iw;
     else
-      wfc_list[iw].restore(iat);
+      idx_Rejected[irej++] = iw;
+
+
+  MultiDiracDeterminant& wfc_leader = wfc_list.getLeader();
+  ParticleSet& p_leader             = p_list.getLeader();
+  const int ndet                    = wfc_leader.getNumDets();
+  const int norb                    = wfc_leader.NumOrbitals;
+  const int nel                     = wfc_leader.NumPtcls;
+  auto& mw_res                      = wfc_leader.mw_res_handle_.getResource();
+
+  const int WorkingIndex = iat - wfc_leader.FirstIndex;
+  assert(WorkingIndex >= 0 && WorkingIndex < wfc_leader.LastIndex - wfc_leader.FirstIndex);
+  assert(p_leader.isSpinor() == wfc_leader.is_spinor_);
+  int handle = 0;
+  for (auto& iacc : idx_Accepted)
+  {
+    auto& wfc = wfc_list.getCastedElement<MultiDiracDeterminant>(iacc);
+    if (wfc.curRatio == ValueType(0))
+    {
+      std::ostringstream msg;
+      msg << "MultiDiracDeterminant::acceptMove curRatio is " << wfc.curRatio << " for walker " << iacc
+          << "! Report a bug." << std::endl;
+      throw std::runtime_error(msg.str());
+    }
+  }
+
+  for (auto& iacc : idx_Accepted)
+  {
+    auto& wfc = wfc_list.getCastedElement<MultiDiracDeterminant>(iacc);
+    wfc.log_value_ref_det_ += convertValueToLog(wfc.curRatio);
+    wfc.curRatio = ValueType(1);
+  }
+
+  // pointers to data for only accepted walkers
+  OffloadVector<ValueType*> psiMinv_temp_acc_deviceptr_list;
+  OffloadVector<ValueType*> psiMinv_acc_deviceptr_list;
+  OffloadVector<ValueType*> psiV_acc_deviceptr_list;
+  OffloadVector<ValueType*> TpsiM_col_acc_deviceptr_list;
+  OffloadVector<ValueType*> psiM_row_acc_deviceptr_list;
+  OffloadVector<ValueType*> new_ratios_to_ref_acc_deviceptr_list;
+  OffloadVector<ValueType*> ratios_to_ref_acc_deviceptr_list;
+
+  OffloadVector<ValueType*> dpsiV_acc_deviceptr_list;
+  OffloadVector<ValueType*> dpsiM_row_acc_deviceptr_list;
+  OffloadVector<ValueType*> d2psiV_acc_deviceptr_list;
+  OffloadVector<ValueType*> d2psiM_row_acc_deviceptr_list;
+
+  Vector<ValueType*> dspin_psiV_acc_ptr_list;
+  Vector<ValueType*> dspin_psiM_row_acc_ptr_list;
+
+  Vector<ValueType*> new_grads_acc_ptr_list;
+  Vector<ValueType*> grads_acc_ptr_list;
+  Vector<ValueType*> new_lapls_acc_ptr_list;
+  Vector<ValueType*> lapls_acc_ptr_list;
+  Vector<ValueType*> new_spingrads_acc_ptr_list;
+  Vector<ValueType*> spingrads_acc_ptr_list;
+
+  /**
+   * some of these are in the mw_resource, and some are not
+   * for the ones that are, get device pointers from the resource collection
+   * for the ones that aren't, get pointers from MultiDiracDeterminant object
+   * TODO: I'm assuming here that all data is already up to date on the device before this function is called
+   */
+
+  // setup device pointer lists
+  switch (wfc_leader.UpdateMode)
+  {
+  default:
```

Comment:
>Before doing a big refactor why not write the missing unit tests. What's the expected outcome of calling this function?

My plan going into this refactor was that it wasn't going to be a huge change, and that I was just planning to group some similar work together (the rejects and accepts) without any other change in behavior.  I don't yet have a complete picture of everything else upstream and downstream from this function and what different cases might need to be handled.

>It seems like this function basically just copies a bunch of state data in response to the isAccepted vector and the internal WorkingIndex state.

This is also my take on what's happening, but the part I'm not sure about (and what I asked about in my comment above) is which state data I can assume to be up-to-date on the host/device before entering this function, and what needs to be up-to-date (and where) after exiting.
This is the only point where I'd planned to possibly deviate from the current functionality (e.g. in the [ORB_PBYP_RATIO case](https://github.com/QMCPACK/qmcpack/blob/65549dbf4715b69fb78030a3100657104dcb30d8/src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.cpp#L427-L445) for accepted moves, we do the [relevant copying (host to host)](https://github.com/QMCPACK/qmcpack/blob/65549dbf4715b69fb78030a3100657104dcb30d8/src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.cpp#L427-L437) to psiMinv, psiM, TpsiM, and ratios_to_ref_, and then we do an [H2D update](https://github.com/QMCPACK/qmcpack/blob/65549dbf4715b69fb78030a3100657104dcb30d8/src/QMCWaveFunctions/Fermion/MultiDiracDeterminant.cpp#L440-L444) of all of those in addition to dpsiM.  If there are cases where all of these (and psiV, psiMinv_temp, and new_ratios_to_ref_) are already up-to-date on the device, and if anything downstream from here will only use those on the device, then it would make sense to just do the copy on the device).